### PR TITLE
DDP-7074 Support for pre-selected answers

### PR DESCRIPTION
**Pull request by YuriyPavlovBI**
https://broadinstitute.atlassian.net/browse/DDP-7074 
Support for pre-selected answers. New attribute is introduced to the pick list option: **_default_**. If question has at least one default option, the default answer will be automatically written right after activity instance creation.

### CODE REVIEW

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/question/PicklistOption.java [hunk 1]
```diff
@@ -37,6 +37,9 @@
     @SerializedName("exclusive")
     private boolean isExclusive;
 
+    @SerializedName("isDefault")
```
Comment (bskinner) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/question/PicklistOption.java]: The API documentation needs to be added for this change if this is a new property being exposed to clients.

**Review by bskinner — CHANGES_REQUESTED** [review]

**Comment by bskinner** [comment]
In the description for the PR, you mention:

> If question has at least one default option, the default answer will be automatically written right after activity instance creation.

Who is doing the creation here? My understanding after looking over the code is that the backend is going to be taking care of creating the default set of answers. If that is the case, why does `isDefault` need to be exposed to clients? If this is not the case, what is the expected flow after a client receives the Activity response?

**Comment by YuriyPavlovBI** [comment]
> In the description for the PR, you mention:
> 
> > If question has at least one default option, the default answer will be automatically written right after activity instance creation.
> 
> Who is doing the creation here? My understanding after looking over the code is that the backend is going to be taking care of creating the default set of answers. If that is the case, why does `isDefault` need to be exposed to clients? If this is not the case, what is the expected flow after a client receives the Activity response?

Yes, you're right, there is no need to expose `isDefault` since backend takes care of it. Removed

File: pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/question/PicklistOption.java (hunk 1 above)
Comment (YuriyPavlovBI) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/question/PicklistOption.java]: Removed

Comment (bskinner) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/model/activity/instance/question/PicklistOption.java]: Per [this comment](https://github.com/broadinstitute/ddp-study-server/pull/1052#issuecomment-960968833), `isDefault` is being removed from the client response so this conversation is moot.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/ActivityInstanceUtil.java [hunk 2]
```diff
@@ -147,4 +161,19 @@ public static boolean computeCanDelete(boolean canDeleteInstance, Boolean canDel
...
+                (questionDef instanceof PicklistQuestionDef) && !((PicklistQuestionDef) questionDef).getDefaultOptions().isEmpty());
+        for (QuestionDef question : questions) {
+            PicklistQuestionDef plQuestion = (PicklistQuestionDef) question;
+            var answer = new PicklistAnswer(null, plQuestion.getStableId(), null,
+                    plQuestion.getDefaultOptions().stream()
+                            .map(plOpt -> new SelectedPicklistOption(plOpt.getStableId()))
+                            .collect(Collectors.toList()));
+            answerDao.createAnswer(operatorId, instanceId, answer, plQuestion);
+        }
+    }
```
Comment (bskinner) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/util/ActivityInstanceUtil.java:178]: Does this handle the case where an option has `isExclusive` set, or when there's a mismatch between the number of default options and any exclusivity or validation requirements?

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistQuestionDao.java [hunk 3]
```diff
@@ -499,23 +499,21 @@ default GroupAndOptionDtos findOrderedGroupAndOptionDtos(long questionId, long t
...
-                if (nestedOptMap.containsKey(parentDto.getId())) {
-                    parentDto.getNestedOptions().addAll(nestedOptMap.get(parentDto.getId()));
-                }
-            });
-        });
+        dtosMap.values().forEach(dto -> dto.getUngroupedOptions().forEach(parentDto -> {
+            if (nestedOptMap.containsKey(parentDto.getId())) {
+                parentDto.getNestedOptions().addAll(nestedOptMap.get(parentDto.getId()));
+            }
+        }));
```
Comment (bskinner) [inline pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistQuestionDao.java]: This looks like an unnecessary change. Unless there is an issue such as a linter failing, let's revert this as the changes make it more difficult to read.

File: pepper-apis/src/main/java/org/broadinstitute/ddp/util/ActivityInstanceUtil.java (hunk 2 above)
Comment (YuriyPavlovBI) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/util/ActivityInstanceUtil.java:178]: No, there is no any checks given the fact that it is a study design time feature in comparison to the runtime stuff like `isExclusive` and validations

File: pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistQuestionDao.java (hunk 3 above)
Comment (YuriyPavlovBI) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistQuestionDao.java]: This change was just for clearing up all the warnings. I changed the shape back to make it looks like before

Comment (YuriyPavlovBI) [reply pepper-apis/src/main/java/org/broadinstitute/ddp/db/dao/PicklistQuestionDao.java]: This change was just for clearing up all the warnings. I changed the shape back to make it looks like before

**Review by bskinner — APPROVED** [review]
👍

**Pull request merged**